src/ros2idl.ne

```diff
@@ -0,0 +1,395 @@
+      msgDefs.push(sub);
+    }
+  });
+  
+  
+  return msgDefs;
+}
+
+function processConstantModule(d) {
+  const moduleName = d[0][2].name;
+  const enclosedConstants = d[0][4];
+  // need to return array here to keep same signature as processComplexModule
+  return [{
+    name: moduleName,
+    definitions: enclosedConstants.flat(1),
+  }];
+}
+
+%}
+
+@lexer lexer
+
+main -> (header:? importDcl:* module:+):+ {% d => {
+  return d[0].flatMap(inner => inner[2].flat());
+}
+%}
+
+header -> %HEADER {% noop %}
+
+# support <import> or "import" includes - just ignored
+importDcl -> "#" "include" (%STRING | "<" %NAME ("/" %NAME):* "." "idl" ">") {% noop %}
+
+# constant modules need to be separate from complex modules since they shouldn't mix
+module  -> (multiAnnotations "module" fieldName "{" (typedefDcl|constantDcl):+ "}" semi) {% processConstantModule %}
+  | (multiAnnotations "module" fieldName "{" (typedefDcl|structWithAnnotation|module):+ "}" semi) {% processComplexModule %}
+
+
+structWithAnnotation -> multiAnnotations struct {% 
+ // default values don't apply to structs so we can just ignore all annotations on structs
+ d => d[1]
+%}
+
+struct -> "struct" fieldName "{" (declaration):+ "}" semi {% d => {
+  const name = d[1].name;
+  const definitions = d[3].flat(2).filter(def => def !== null);
+  return {
+    name,
+    definitions,
+  };
+} %}
+
+typedefDcl -> typedefWithAnnotations semi {% noop %}
+
+typedefWithAnnotations -> multiAnnotations (
+   typedef allTypes fieldName arrayLength
+ | typedef allTypes fieldName
+ | typedef sequenceType fieldName
+) {% d => {
+  const [{name}] = d[1].splice(2, 1);
+  const def = extend(d.flat(1));
+  typedefMap[name] = extend(d.flat(1));
+  return null;
+} %}
+
+typedef -> "typedef" {% noop %}
+constantDcl -> multiAnnotations constType semi {% d => d[1] %}
+declaration -> fieldWithAnnotation semi {% d => d[0] %}
+
+fieldWithAnnotation -> multiAnnotations fieldDcl {% d=> {
+  let possibleAnnotations = [];
+  if(d[0]) {
+    possibleAnnotations = d[0];
+  }
+  const fields = d[1];
+  const finalDefs = fields.map((def) => extend([...possibleAnnotations, def]));
+  return finalDefs;
+} %}
+
+fieldDcl -> (
+     allTypes  multiFieldNames arrayLength
+   | allTypes multiFieldNames
+   | sequenceType multiFieldNames
+ ) {% (d) => {
+  const names = d[0].splice(1, 1)[0];
+  // create a definition for each name
+  const defs = names.map((nameObj) => extend([...d[0], nameObj]));
+  return defs
+} %}
+
+multiFieldNames -> fieldName ("," fieldName):* {%
+ d => {
+   const fieldNames = d.flat(2).filter( d => d !== null && d.name);
+   return fieldNames;
+ } %}
+   
+multiAnnotations -> annotation:* {%
+  d => {
+    return d[0] ? d[0].filter(d => d !== null) : null;
+  }
+%}
+
+annotation -> at %NAME ("(" multiAnnotationParams ")"):? {% d => {
+  const paramsMap = d[2] ? d[2][1] : {};
+  if(d[1].value === "default") {
+    const defaultValue = paramsMap.value;
+    return {defaultValue};
+  }
```

Comment:
I would suggest moving this out of the grammar and into a postprocessing step, so the grammar remains a fairly simple AST generator